README.md
[![Donate](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=3HHNV9W5RGDPJ&source=url)

# hsl_colorpicker

A HSL Color Picker for Flutter (Currently only supports Android and iOS)

| Example app             | Example in other project|
| ------------------------| ------------------------|
| ![](images/example.gif) | ![](images/example.png) |


## How to use it

### Installation

You just need to add `hsl_colorpicker` as a [dependency in your pubspec.yaml file](https://flutter.io/using-packages/).

```yaml
dependencies:
  hsl_colorpicker: ^1.0.1
```
Then run the flutter tooling:

```yaml
flutter packages get
```

### Usage
Put the `HSLColorPicker` widget as child of some other widget

```dart
import 'package:hsl_colorpicker/HSLColorPicker.dart';


HSLColorPicker(
  onChanged: (HSLColor color) {
    // Handle color changes
   },
),
```

### API

```dart
  /// Called during a drag when the user is selecting a color.
  /// Is the only required parameter.
  ///
  /// This callback called with latest HSLColor that user selected.
  final ValueChanged<HSLColor> onChanged;
  
  
  /// Initial color for picker.
  /// [onChanged] callback won't be called with initial value.
  ///
  /// Default value is Blue accent.
  final Color initialColor;


  /// The size of the ColorPicker.
  /// The size of the thumb is not included in the size of the ColorPicker,
  /// If the size of the thumb is greater than the strokeSize of the hue circle, it will be drawn beyond the size.
  ///
  /// Default value is 200 x 200.
  final double size;


  /// The width of hue color circle and saturation and lightness archs.
  ///
  /// Default value is relative to the size of the colorpicker.
  double strokeWidth;


  /// The size of each thumb in the picker.
  ///
  /// Default value is relative to the size of the colorpicker.
  double thumbSize;
  

  /// The size of each thumb strokes.
  ///
  /// Default value is relative to the size of the colorpicker.
  double thumbStrokeSize;
  

  /// Show or not the middle circle that indicates the selected color of the colorpicker.
  ///
  /// Default value is true.
  bool showCenterColorIndicator;
  

  /// The size of the middle circle that indicates the selected color of the colorpicker.
  /// The distance between the hue circle and the archs of lightness and saturation is determined by the size of the colorpicker and the   /// size of the middle circle that indicates the selected color.
  ///
  /// Default value is relative to the size of the colorpicker.
  bool centerColorIndicatorSize;  
```

### Example

```dart
import 'package:flutter/material.dart';
import 'package:hsl_colorpicker/HSLColorPicker.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: ColorPickerPage(),
      debugShowCheckedModeBanner: false,
    );
  }
}

class ColorPickerPage extends StatefulWidget {
  @override
  _ColorPickerPageState createState() => _ColorPickerPageState();
}

class _ColorPickerPageState extends State<ColorPickerPage> {
  HSLColor hslColor = HSLColor.fromColor(Colors.blue);
  Color color = Colors.blue;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.white,
      appBar: AppBar(
        backgroundColor: color,
        title: Stack(
          children: <Widget>[
            Text(
              "HSL COLOR PICKER: ${"#" + color.toString().split("0x")[1].toUpperCase().replaceFirst("FF", "").replaceAll(")", "")}",
              style: TextStyle(
                fontSize: 16,
                foreground: Paint()
                  ..style = PaintingStyle.stroke
                  ..strokeWidth = 1
                  ..color = Colors.grey,
              ),
            ),
            Text(
              "HSL COLOR PICKER: ${"#" + color.toString().split("0x")[1].toUpperCase().replaceFirst("FF", "").replaceAll(")", "")}",
              style: TextStyle(
                color: Colors.white,
                fontSize: 16,
              ),
            ),
          ],
        ),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            HSLColorPicker(
              onChanged: (colorSelected) {
                setState(() {
                  hslColor = colorSelected;
                  color = colorSelected.toColor();
                });
              },
              size: 200,
              strokeSize: 5,
              thumbSize: 9,
              thumbStrokeSize: 3,
              showCenterColorIndicator: true,
              centerColorIndicatorSize: 80,
              initialColor: Colors.blueAccent,
            ),
          ],
        ),
      ),
    );
  }
}
```


CHANGELOG.md
## 1.0.2
* Changes on Readme

## 1.0.1
* Initial Release


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:hslcolorpicker/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/ColorPicker.dart
import 'package:flutter/material.dart';
import 'package:hsl_colorpicker/HSLColorPicker.dart';

class ColorPicker extends StatefulWidget {
  final Color mColor;
  final double anchura;
  final double altura;
  final double radioClip;

  const ColorPicker(
      {Key key, this.mColor, this.anchura, this.altura, this.radioClip})
      : super(key: key);

  @override
  _ColorPickerState createState() => _ColorPickerState();
}

class _ColorPickerState extends State<ColorPicker> {
  HSLColor color;
  HSLColor colorHex;
  double luminosidad = 1;
  double luminosidadHex = 1;

  @override
  void initState() {
    color = HSLColor.fromColor(widget.mColor);
    colorHex = HSLColor.fromColor(widget.mColor);
    super.initState();
  }

  void onChanged(HSLColor color) {
    this.color = color;

    if (color.lightness > 0.8) {
      double variable = color.lightness - 0.8;
      luminosidad = 1 - (variable * 5);
    } else {
      luminosidad = 1;
    }
  }

  @override
  Widget build(BuildContext context) {
    return Center(
      child: Stack(
        children: <Widget>[
          Padding(
            padding: EdgeInsets.only(bottom: 6),
            child: Center(
              child: Container(
                width: 230,
                height: 230,
                child: HSLColorPicker(
                  initialColor: Colors.blue,
                  onChanged: (value) {
                    super.setState(() {
                      this.onChanged(value);
                    });
                  },
                ),
              ),
            ),
          ),
          Center(
            child: Container(
              height: 80,
              width: 80,
              child: Material(
                borderRadius: BorderRadius.circular(64),
                color: color.toColor(),
                child: Center(
                  child: Text(
                    "#" +
                        color
                            .toColor()
                            .toString()
                            .split("0x")[1]
                            .toUpperCase()
                            .replaceFirst("FF", "")
                            .replaceAll(")", ""),
                    style: TextStyle(
                        color: HSLColor.fromColor(Colors.grey)
                            .withLightness(luminosidad)
                            .toColor()),
                  ),
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }
}

int hexToInt(String hex) {
  int val = 0;
  int len = hex.length;
  for (int i = 0; i < len; i++) {
    int hexDigit = hex.codeUnitAt(i);
    if (hexDigit >= 48 && hexDigit <= 57) {
      val += (hexDigit - 48) * (1 << (4 * (len - 1 - i)));
    } else if (hexDigit >= 65 && hexDigit <= 70) {
      // A..F
      val += (hexDigit - 55) * (1 << (4 * (len - 1 - i)));
    } else if (hexDigit >= 97 && hexDigit <= 102) {
      // a..f
      val += (hexDigit - 87) * (1 << (4 * (len - 1 - i)));
    } else {
      throw new FormatException("Invalid hexadecimal value");
    }
  }
  return val;
}


lib/HSLColorPicker.dart
library hslcolorpicker;

import "dart:math" as Math;

import 'package:flutter/material.dart';

/// SIZE OF A SINGLE PIXEL ON THE SCREEN
double pixelSize;

class HSLColorPicker extends StatefulWidget {
  final ValueChanged<HSLColor> onChanged;
  final Color initialColor;
  final double size;
  double strokeWidth;
  double thumbSize;
  double thumbStrokeSize;
  bool showCenterColorIndicator;
  double centerColorIndicatorSize;

  HSLColorPicker({
    Key key,
    @required this.onChanged,
    this.initialColor = Colors.blueAccent,
    this.size = 200,
    this.centerColorIndicatorSize,
    this.showCenterColorIndicator = true,
    this.strokeWidth,
    this.thumbSize,
    this.thumbStrokeSize,
  })  : assert(onChanged != null),
        super(key: key);

  @override
  _HSLColorPickerState createState() => new _HSLColorPickerState();
}

class _HSLColorPickerState extends State<HSLColorPicker> {
  HSLColor color;
  final GlobalKey colorPickerKey = GlobalKey();

  /// Values
  double hue;
  double lightness;
  double saturation;

  /// Controls
  bool controles;
  bool controlSaturacion;

  @override
  void didChangeDependencies() {
    pixelSize = 1.0 / MediaQuery.of(context).devicePixelRatio;
    super.didChangeDependencies();
  }

  @override
  void initState() {
    color = HSLColor.fromColor(widget.initialColor);
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    hue = color.hue;
    lightness = color.lightness;
    saturation = color.saturation;

    if (widget.strokeWidth == null) {
      widget.strokeWidth = widget.size * .027;
    }
    if (widget.thumbSize == null) {
      widget.thumbSize = widget.size * .043;
    }
    if (widget.thumbStrokeSize == null) {
      widget.thumbStrokeSize = widget.size * .015;
    }
    if (widget.centerColorIndicatorSize == null) {
      widget.centerColorIndicatorSize = widget.size / 2.6;
    }

    return GestureDetector(
      onPanStart: (details) => this.onPanStart(details.globalPosition),
      onPanUpdate: (details) => this.onPanUpdate(details.globalPosition),
      child: Container(
        key: this.colorPickerKey,
        child: CustomPaint(
          size: Size(widget.size, widget.size),
          painter: new _WheelPainter(
              color: this.color,
              strokeSize: widget.strokeWidth,
              showCenterColorIndicator: widget.showCenterColorIndicator,
              thumbSize: widget.thumbSize,
              centerColorIndicatorSize: widget.centerColorIndicatorSize,
              thumbStrokeSize: widget.thumbStrokeSize),
        ),
      ),
    );
  }

  Offset getOffset(Offset ratio) {
    RenderBox renderBox = this.colorPickerKey.currentContext.findRenderObject();
    Offset startPosition = renderBox.localToGlobal(Offset.zero);
    return ratio - startPosition;
  }

  Size getSize() {
    RenderBox renderBox = this.colorPickerKey.currentContext.findRenderObject();
    return renderBox.size;
  }

  void onPanStart(Offset offset) {
    RenderBox renderBox = this.colorPickerKey.currentContext.findRenderObject();
    Size size = renderBox.size;

    Offset startPosition = renderBox.globalToLocal(offset);
    Offset center = Offset(size.width / 2, size.height / 2);

    Offset vector = startPosition - center;

    final maxRadioControles = _WheelPainter.radio(size, widget.strokeWidth,
            widget.thumbSize, widget.thumbStrokeSize) -
        10;
    final hip = Math.sqrt(Math.pow(vector.dx, 2) + Math.pow(vector.dy, 2));

    if (hip < maxRadioControles) {
      controles = true;
      var gradosCircunferencia =
          Math.atan2(vector.dy, vector.dx) * 180 / Math.pi;

      if (gradosCircunferencia < 0) {
        gradosCircunferencia = 180 + 180 - gradosCircunferencia.abs();
      }

      controlSaturacion =
          (gradosCircunferencia > 90 && gradosCircunferencia < 270);
    } else {
      controles = false;
    }

    _actualizar(vector);
  }

  void onPanUpdate(Offset offset) {
    RenderBox renderBox = this.colorPickerKey.currentContext.findRenderObject();
    Size size = renderBox.size;

    Offset startPosition = renderBox.globalToLocal(offset);
    Offset center = Offset(size.width / 2, size.height / 2);

    Offset vector = startPosition - center;
    _actualizar(vector);
  }

  void _actualizar(Offset vector) {
    setState(() {
      var gradosCircunferencia =
          Math.atan2(vector.dy, vector.dx) * 180 / Math.pi;

      if (gradosCircunferencia < 0) {
        gradosCircunferencia = 180 + 180 - gradosCircunferencia.abs();
      }

      if (controles) {
        if (controlSaturacion) {
          if (gradosCircunferencia < 270 && gradosCircunferencia > 90) {
            var gradosSaturacion = gradosCircunferencia - 90;

            if (gradosSaturacion > 180 - _WheelPainter.apertura) {
              gradosSaturacion = 180 - _WheelPainter.apertura;
            }
            if (gradosSaturacion < _WheelPainter.apertura) {
              gradosSaturacion = _WheelPainter.apertura;
            }

            saturation = (1.0 /
                _WheelPainter.gradosArco *
                (gradosSaturacion - _WheelPainter.apertura));
          }
        } else {
          if (gradosCircunferencia > 270 || gradosCircunferencia < 90) {
            var gradosLuminosidad;
            if (gradosCircunferencia > 270) {
              gradosLuminosidad = gradosCircunferencia - 270;
            } else {
              gradosLuminosidad = gradosCircunferencia + 90;
            }

            if (gradosLuminosidad > 180 - _WheelPainter.apertura) {
              gradosLuminosidad = 180 - _WheelPainter.apertura;
            }
            if (gradosLuminosidad < _WheelPainter.apertura) {
              gradosLuminosidad = _WheelPainter.apertura;
            }

            lightness = (1.0 /
                _WheelPainter.gradosArco *
                (gradosLuminosidad - _WheelPainter.apertura));
          }
        }
      } else {
        hue = gradosCircunferencia;
      }

      color = HSLColor.fromAHSL(1, hue, saturation, lightness);
      super.widget.onChanged(HSLColor.fromAHSL(1, hue, saturation, lightness));
    });
  }
}

class Wheel {
  static Offset hueToVector(double h, double radio, Offset center) =>
      new Offset(
          Math.cos(h) * radio + center.dx, Math.sin(h) * radio + center.dy);
}

class _WheelPainter extends CustomPainter {
  static double radio(Size size, double strokeSize, double thumbSize,
          double thumbStrokeSize) =>
      Math.min(size.width, size.height).toDouble() / 2 -
      Math.max(strokeSize, thumbSize + (thumbStrokeSize / 2));

  final HSLColor color;
  final double strokeSize;
  final bool showCenterColorIndicator;
  final double thumbSize;
  final double thumbStrokeSize;
  final double centerColorIndicatorSize;

  static final apertura = 15.0;
  static final gradosArco = 180 - apertura * 2;
  static double space(Size size, double centerColorIndicatorSize,
          double strokeSize, double thumbSize, double thumbStrokeSize) =>
      (Math.min(size.width, size.height).toDouble() -
          centerColorIndicatorSize -
          Math.max(strokeSize, thumbSize + (thumbStrokeSize / 2))) /
      4;

  _WheelPainter({
    Key key,
    this.color,
    this.showCenterColorIndicator,
    this.centerColorIndicatorSize,
    this.strokeSize,
    this.thumbSize,
    this.thumbStrokeSize,
  }) : super();

  @override
  void paint(Canvas canvas, Size size) {
    Offset center = new Offset(size.width / 2, size.height / 2);

    Offset centerOffset = Offset(0, 0);
    double radio =
        _WheelPainter.radio(size, strokeSize, thumbStrokeSize, thumbStrokeSize);

    canvas.translate(center.dx, center.dy);

    /// center indicator color circle
    if (showCenterColorIndicator) {
      canvas.drawCircle(
          centerOffset,
          (centerColorIndicatorSize / 2) + pixelSize,
          Paint()
            ..style = PaintingStyle.fill
            ..color = Colors.grey);

      canvas.drawCircle(
          centerOffset,
          centerColorIndicatorSize / 2,
          Paint()
            ..style = PaintingStyle.fill
            ..color = color.toColor());
    }

    /// hue circle
    canvas.drawCircle(
      centerOffset,
      radio,
      Paint()
        ..style = PaintingStyle.stroke
        ..strokeWidth = strokeSize
        ..shader = SweepGradient(
          colors: [
            HSLColor.fromAHSL(1, 0, 1, .5).toColor(),
            HSLColor.fromAHSL(1, 60, 1, .5).toColor(),
            HSLColor.fromAHSL(1, 120, 1, .5).toColor(),
            HSLColor.fromAHSL(1, 180, 1, .5).toColor(),
            HSLColor.fromAHSL(1, 240, 1, .5).toColor(),
            HSLColor.fromAHSL(1, 300, 1, .5).toColor(),
            HSLColor.fromAHSL(1, 0, 1, .5).toColor(),
          ],
        ).createShader(
          Rect.fromCircle(center: centerOffset, radius: radio),
        ),
    );

    canvas.rotate(Math.pi / 180 * 270);

    canvas.drawArc(
        Rect.fromCircle(
            center: Offset(0, 3),
            radius: radio -
                space(size, centerColorIndicatorSize, strokeSize,
                    thumbStrokeSize, thumbStrokeSize)),
        Math.pi / 180 * apertura,
        Math.pi / 180 * gradosArco,
        false,
        Paint()
          ..style = PaintingStyle.stroke
          ..strokeCap = StrokeCap.round
          ..strokeWidth = strokeSize + pixelSize * 2
          ..color = Colors.grey);

    /// lightness arch
    canvas.drawArc(
      Rect.fromCircle(
          center: Offset(0, 3),
          radius: radio -
              space(size, centerColorIndicatorSize, strokeSize, thumbStrokeSize,
                  thumbStrokeSize)),
      Math.pi / 180 * apertura,
      Math.pi / 180 * gradosArco,
      false,
      Paint()
        ..style = PaintingStyle.stroke
        ..strokeCap = StrokeCap.round
        ..strokeWidth = strokeSize
        ..shader = SweepGradient(
          startAngle: Math.pi / 180 * apertura,
          endAngle: Math.pi / 180 * gradosArco,
          colors: [
            HSLColor.fromAHSL(
              1,
              color.hue,
              color.saturation,
              0,
            ).toColor(),
            HSLColor.fromAHSL(
              1,
              color.hue,
              color.saturation,
              .5,
            ).toColor(),
            HSLColor.fromAHSL(
              1,
              color.hue,
              color.saturation,
              1,
            ).toColor(),
          ],
        ).createShader(
          Rect.fromCircle(
              center: Offset(0, 3),
              radius: radio -
                  space(size, centerColorIndicatorSize, strokeSize,
                      thumbStrokeSize, thumbStrokeSize)),
        ),
    );
    canvas.rotate(-Math.pi / 180 * 270);

    canvas.rotate(Math.pi / 180 * 90);

    /// saturation arch
    canvas.drawArc(
        Rect.fromCircle(
            center: Offset(0, 3),
            radius: radio -
                space(size, centerColorIndicatorSize, strokeSize,
                    thumbStrokeSize, thumbStrokeSize)),
        Math.pi / 180 * apertura,
        Math.pi / 180 * gradosArco,
        false,
        Paint()
          ..style = PaintingStyle.stroke
          ..strokeCap = StrokeCap.round
          ..strokeWidth = strokeSize + pixelSize * 2
          ..color = Colors.grey);

    canvas.drawArc(
      Rect.fromCircle(
          center: Offset(0, 3),
          radius: radio -
              space(size, centerColorIndicatorSize, strokeSize, thumbStrokeSize,
                  thumbStrokeSize)),
      Math.pi / 180 * apertura,
      Math.pi / 180 * gradosArco,
      false,
      Paint()
        ..style = PaintingStyle.stroke
        ..strokeCap = StrokeCap.round
        ..strokeWidth = strokeSize
        ..shader = SweepGradient(
          startAngle: Math.pi / 180 * apertura,
          endAngle: Math.pi / 180 * gradosArco,
          colors: [
            HSLColor.fromAHSL(1, color.hue, 0, color.lightness).toColor(),
            HSLColor.fromAHSL(1, color.hue, 1, color.lightness).toColor(),
          ],
        ).createShader(
          Rect.fromCircle(
              center: Offset(0, 3),
              radius: radio -
                  space(size, centerColorIndicatorSize, strokeSize,
                      thumbStrokeSize, thumbStrokeSize)),
        ),
    );

    canvas.rotate(-Math.pi / 180 * 90);

    /// Thumb hue picker
    final Paint paintColor = new Paint()
      ..color = color.toColor()
      ..style = PaintingStyle.fill;
    final Paint paintWhite = new Paint()
      ..color = Colors.white
      ..strokeWidth = thumbStrokeSize
      ..style = PaintingStyle.stroke;
    final Paint paintBlack = new Paint()
      ..color = Colors.grey
      ..strokeWidth = thumbStrokeSize + 2
      ..style = PaintingStyle.stroke;

    Offset wheel = Wheel.hueToVector(
        ((color.hue + 360.0) * Math.pi / 180.0), radio, centerOffset);
    canvas.drawCircle(wheel, thumbSize, paintBlack);
    canvas.drawCircle(wheel, thumbSize, paintWhite);
    canvas.drawCircle(wheel, thumbSize, paintColor);

    /// Thumb saturacion picker
    final Paint paintSaturacionColor = new Paint()
      ..color = color.toColor()
      ..style = PaintingStyle.fill;
    final Paint paintSaturacionWhite = new Paint()
      ..color = Colors.white
      ..strokeWidth = thumbStrokeSize
      ..style = PaintingStyle.stroke;
    final Paint paintSaturacionBlack = new Paint()
      ..color = Colors.grey
      ..strokeWidth = thumbStrokeSize + 2
      ..style = PaintingStyle.stroke;

    Offset saturacionWheel = Wheel.hueToVector(
        ((color.saturation * gradosArco + apertura + 90) * Math.pi / 180.0),
        radio -
            space(size, centerColorIndicatorSize, strokeSize, thumbStrokeSize,
                thumbStrokeSize),
        centerOffset);
    canvas.drawCircle(
        saturacionWheel.translate(-3, 0), thumbSize, paintSaturacionBlack);
    canvas.drawCircle(
        saturacionWheel.translate(-3, 0), thumbSize, paintSaturacionWhite);
    canvas.drawCircle(
        saturacionWheel.translate(-3, 0), thumbSize, paintSaturacionColor);

    /// Thumb Lightness picker
    final Paint paintLuminosidadColor = new Paint()
      ..color = color.toColor()
      ..style = PaintingStyle.fill;
    final Paint paintLuminosidadWhite = new Paint()
      ..color = Colors.white
      ..strokeWidth = thumbStrokeSize
      ..style = PaintingStyle.stroke;
    final Paint paintLuminosidadBlack = new Paint()
      ..color = Colors.grey
      ..strokeWidth = thumbStrokeSize + 2
      ..style = PaintingStyle.stroke;

    Offset luminosidadWheel = Wheel.hueToVector(
        ((color.lightness * gradosArco + apertura - 90) * Math.pi / 180.0),
        radio -
            space(size, centerColorIndicatorSize, strokeSize, thumbStrokeSize,
                thumbStrokeSize),
        centerOffset);
    canvas.drawCircle(
        luminosidadWheel.translate(3, 0), thumbSize, paintLuminosidadBlack);
    canvas.drawCircle(
        luminosidadWheel.translate(3, 0), thumbSize, paintLuminosidadWhite);
    canvas.drawCircle(
        luminosidadWheel.translate(3, 0), thumbSize, paintLuminosidadColor);
  }

  @override
  bool shouldRepaint(_WheelPainter other) => true;
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:hsl_colorpicker/HSLColorPicker.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: ColorPickerPage(),
      debugShowCheckedModeBanner: false,
    );
  }
}

class ColorPickerPage extends StatefulWidget {
  @override
  _ColorPickerPageState createState() => _ColorPickerPageState();
}

class _ColorPickerPageState extends State<ColorPickerPage> {
  HSLColor hslColor = HSLColor.fromColor(Colors.blue);
  Color color = Colors.blue;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.white,
      appBar: AppBar(
        backgroundColor: color,
        title: Stack(
          children: <Widget>[
            Text(
              "HSL COLOR PICKER: ${"#" + color.toString().split("0x")[1].toUpperCase().replaceFirst("FF", "").replaceAll(")", "")}",
              style: TextStyle(
                fontSize: 16,
                foreground: Paint()
                  ..style = PaintingStyle.stroke
                  ..strokeWidth = 1
                  ..color = Colors.grey,
              ),
            ),
            Text(
              "HSL COLOR PICKER: ${"#" + color.toString().split("0x")[1].toUpperCase().replaceFirst("FF", "").replaceAll(")", "")}",
              style: TextStyle(
                color: Colors.white,
                fontSize: 16,
              ),
            ),
          ],
        ),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            HSLColorPicker(
              onChanged: (colorSelected) {
                print(colorSelected);
                setState(() {
                  hslColor = colorSelected;
                  color = colorSelected.toColor();
                });
              },
              size: 200,
              strokeWidth: 5,
              thumbSize: 9,
              thumbStrokeSize: 3,
              showCenterColorIndicator: true,
              centerColorIndicatorSize: 80,
              initialColor: Colors.blueAccent,
            ),
          ],
        ),
      ),
    );
  }
}
